AzureIoTHub.agent.lib.nut

```diff
@@ -629,513 +615,840 @@ class AzureIoTHub {
+            _topics.msgSend <- format("devices/%s/messages/events/", deviceId);
+            _topics.msgRecv <- format("devices/%s/messages/devicebound/", deviceId);
+            _topics.twinRecv <- "$iothub/twin/res/";
+            _topics.twinGet <- "$iothub/twin/GET/";
+            _topics.twinUpd <- "$iothub/twin/PATCH/properties/reported/";
+            _topics.twinNotif <- "$iothub/twin/PATCH/properties/desired/";
+            _topics.dMethodNotif <- "$iothub/methods/POST/";
+            _topics.dMethodResp <- "$iothub/methods/res/";
         }
 
-        function _connectionCallback(error) {
-            _connecting = false;
-            if (_handlers.onConnected) {
-                imp.wakeup(0, function() {
-                    _handlers.onConnected(error);
-                }.bindenv(this));
+        function _onConnect(err) {
+            if (err == 0) {
+                _isConnected = true;
+                _isDisconnected = false;
+            }
+            _isConnecting = false;
+            // TODO: What should we pass to _onConnectCb?
+            _onConnectCb(err);
+        }
+
+        function _onDisconnect() {
+            _log("Disconnected!");
+            local reason = _isDisconnecting ? 0 : AZURE_IOT_CLIENT_ERROR_GENERAL;
+            _cleanup();
+            _onDisconnectCb && _onDisconnectCb(reason);
+        }
+
+        function _onMessage(msg) {
+            local message = null;
+            local topic = null;
+            try {
+                message = msg["message"];
+                topic = msg["topic"];
+                _log(format("_onMessage: topic=%s | body=%s", topic, message));
+            } catch (e) {
+                _logError("Could not read message: " + e);
+                return;
+            }
+
+            // Cloud-to-device message received
+            if (topic.find(_topics.msgRecv) != null) {
+                _handleCloudToDevMsg(message, topic);
+            // Desired properties were updated
+            } else if (topic.find(_topics.twinNotif) != null) {
+                _handleDesPropsMsg(message, topic);
+            // Twin request result received
+            } else if (topic.find(_topics.twinRecv) != null) {
+                _handleTwinResponse(message, topic);
+            // Direct method called
+            } else if (topic.find(_topics.dMethodNotif) != null) {
+                _handleDirMethodMsg(message, topic);
             }
         }
 
-        function _resetConnectionTables() {
-            _sessions = _getDefaultConnectionTable();
-            _transfers = _getDefaultConnectionTable();
-            _senders = _getDefaultConnectionTable();
-            _receivers = _getDefaultConnectionTable();
-            _msgQueue = [];
-            _handlers = {};
+        function _handleCloudToDevMsg(message, topic) {
+            local props = {};
+            local splittedTopic = split(topic, "/");
+            if (splittedTopic[splittedTopic.len() - 1] != "devicebound") {
```

Comment:
is indexOutOfBounds exception possible here?

Reply:
It is possible only if splittedTopic.len() is 0 which means the topic doesn't contain any subtopics. But we never subscribe to such topics. May the topic be corrupted for some reason? In that case yes, exception is possible.

Reply:
Done